app/usecase/usecase.go

```diff
@@ -0,0 +1,72 @@
+package usecase
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/byliuyang/app/fw"
+	"github.com/byliuyang/kgs/app/adapter/message"
+	"github.com/byliuyang/kgs/app/entity"
+	"github.com/byliuyang/kgs/app/usecase/keys"
+	"github.com/byliuyang/kgs/app/usecase/notification"
+)
+
+type UseCase struct {
+	logger   fw.Logger
+	template fw.Template
+	producer keys.Producer
+	consumer keys.Consumer
+	notifier notification.Notifier
+}
+
+func (u UseCase) PopulateKey(keyLength uint, requesterEmail string) {
+	startAt := time.Now()
+	err := u.producer.Produce(keyLength)
+	if err != nil {
+		u.logger.Error(err)
+		return
+	}
+
+	timeElapsed := time.Since(startAt)
+	msg, err := message.NewKeyGenSucceedMessage(u.template, timeElapsed)
+	if err != nil {
+		u.logger.Error(err)
+		return
+	}
+
+	requester := entity.Requester{
+		Name:  "",
+		Email: requesterEmail,
+	}
+	err = u.notifier.NotifyRequester(msg, requester)
```

Comment:
I have a proposition here...
It seems to me that it is an ideal place for `Event Dispatcher` pattern (We don't have any guarantee `email, SMS, push notification` to be delivered). It will give us more flexibility and a more natural way to inform that something has changed. It can be overwritten in this way:

```go
func (u UseCase) PopulateKey(keyLength uint, requesterEmail string) {
	startAt := time.Now()
	
        if err := u.producer.Produce(keyLength); err != nil {
		u.logger.Error(err)
		return
	}

        u.eventDispatcher.Publish(PopulatedKeyEvent{
        	requester: ...,
        	timeElapsed: time.Since(startAt),
        })

        u.logger.Info("Finish populating keys")
}

```

As an implementation, we can use this [package](https://github.com/asaskevich/EventBus). What do you think?

Reply:
Good work! I like the way you separated the business logic.

Reply:
@alldroll I wonder which one should we choose between message queue or pub sub (https://streaml.io/resources/tutorials/concepts/messaging-and-queuing). The second question is, whether we should make event system as implementation detail of the notifier. I am more interested in building a separate message queue or pub-sub service so that all our microservices can enjoy the benefits. Btw, I also wrote an eventbus (https://github.com/byliuyang/eventbus/blob/master/eventbus.go) and used it here (https://github.com/byliuyang/app/blob/master/tool/sample.go)!

Reply:
This PR is mainly about refactoring rather than adding new functionality. I can add a TODO or open a new issue so that we can discuss in detail about the design decisions.

Reply:
I think we can use 2 different approaches together (here I mean `message queue or pub sub` and `event system`). In my opinion, it gives more flexibility

```go
...
eventDispatcher.Publish(authenticationEvent)

...
func (e EventListener) OnAuth(event Event) error {
     e.messageQueue.Push(Message{...})
}
```

What do you think?

Reply:
@alldroll I see. So event based systems are async. RPC systems are sync. I wonder whether if we completely adopt async approach, when should we register all the event listeners?

Reply:
@byliuyang I think we should run event listeners on the app start, as a separate `goroutine`

Reply:
@alldroll is this what you are thinking about? https://gameprogrammingpatterns.com/event-queue.html

Reply:
@byliuyang Yes, definitely this:)

Reply:
@alldroll https://github.com/short-d/kgs/issues/51